### DISCUSSION TITLE
Convert and Mogrify in Parallel

### DISCUSSION THREAD

**Discussion by PorkeyThePine** [ImageMagick/ImageMagick#3775] [Help] [answered]
I have the occasional need to convert large batches of (regular-sized) images between formats, and I've been having trouble getting ImageMagick to do it _efficiently_.

Running Windows, I have been able to use batch scripts to get mogrify to convert between image formats;
`mogrify -format jpg -path "C:\path\to\converted\files" "C:\source\directory\*.png" -quality 91`

As well as recursively, in case of subfolders;
`FOR /R %f IN (*.png) DO magick mogrify -format tiff -verbose -compress ZIP "%f"
del /s *.png`

However I am running into a problem where imagemagick is processing only one image at a time. This shouldn't technically be an issue, but imagemagick is also doing the conversion on a single thread. OpenMP 2.0 is listed in magick's features, so it should be enabled?

I have a 3950x, 64GB of RAM, two GPUs, and want to be able to put that speed to use. Whether by converting the images multithreaded one at a time, or multiple images at once on their own threads; whichever is faster.
How can I do this with imagemagick, and is it even the best tool for this kind of use?

**Comment by PorkeyThePine** [comment]
`FORFILES /m *.png /s /c "cmd /c start /min magick mogrify -format tiff -compress ZIP @path"`
Works, but will instantly spawn X number of processes where X is the number of images in the **entire** directory tree, **recursively.**

If there would be some way of easily limiting the number of running processes, only starting new ones as old ones finish and terminate, this would effectively accomplish what I (think) I want.

**Comment by snibgo** [comment]
IM has built-in multi-threading for most if its image-processing operations. But you are not doing any processing, just reading and writing.

Do you have a rotating disk, or SSD? If rotating, I expect disk access is the bottleneck, and multi-processing many images at the same time probably has no advantage. But you can test for this.

In Windows, taskmgr will indicate where your bottleneck is.

It is possible to limit the number of running processes, but not simple.

A simpler way to implement multi-processing is:

1. Decide how many processes you want to run simultaneously, such as 8.

2. In a script, loop through your input files. For each file, write the appropriate "magick" command to one of 8 script files. As this does hardly any processing per file, it is fast.

3. Start each of those 8 script files.

**Comment by PorkeyThePine** [comment]
I've just run a quick test on a folder containing 21,275 .png files (1920x1080 screenshots), converting them to .jpg in a different folder.
Canceled the test after about a minute as that folder isn't actually "ready" for full conversion yet.
Magick.exe was running along at 9-12% CPU usage in Task Manager's Details tab.
Disk read/write speed was in the 40MB/s range.
Disk activity was 15-25%.
Disk I/O queue in Resource Monitor was 0.1-0.45, with only occasional spikes (every 5-10 seconds) above 1.0 (highest 1.15).
Test disk is a Seagate FireCuda ST2000DX002. There is no way the test files are already on the SSD cache.

A disk bottleneck would certainly make a lot of sense, but task manager seems to be telling me this is a software bottleneck?

I ran an earlier test, converting a folder (containing subfolders) totaling 1.32GB in 68 .png files, converting them to .tiff files in the same folder structure (converted .tiffs ended up beside each .png file).
The operation took a few seconds; I did not measure exact time to completion but maybe ~10 seconds or so?
During this operation, I recall magick.exe using 3% CPU (1 core).

Using the command
FORFILES /m *.png /s /c "cmd /c start /min magick mogrify -format tiff -compress ZIP @path"
on the same 68-file directory had the operation complete more-or-less instantly (with a single 100% CPU _spike_).
This would not be the ideal command to use on the larger .png-to-.jpg directory...

I don't like it, but so far it does seem that multiprocessing is useful in this type of use case.
Unfortunately I've never had to do scripting as advanced as what you're suggesting. Looks like I've got some learnin' to do unless someone can help and point me in the right direction.

**Comment by snibgo** [comment, marked as answer]
> Magick.exe was running along at 9-12% CPU usage in Task Manager's Details tab.

How many cores do you have? That would be a typical usage on my laptop, which has 8 cores, saturating a single core. IM reading and writing files is single-threaded.

So, assuming you have 8 cores, expect 8 parallel processes to saturate the CPU.

Also watch for memory usage, but I doubt that is a problem as your images are small (2 MP each).

Your disk usage is quite high, so that might limit sensible parallel processes to 4. Beyond that, there may be no improvement.

"magick mogrify" is useful when converting a number of files. But I suggest you do them one at a time, so use just "magick".

So, each of your 4 script files will contain lines like this:
```
magick file1.png outdir\file1.jpg
magick file2.png outdir\file2.jpg
magick file3.png outdir\file3.jpg
```
... and then you "cmd /c start /min" each of those 4 script files.

To create each script file, I would use a "for" loop, that uses "echo" to write the magick command to one of the four script files. It could cycle through the script files, so the first image file went to the first script file, the second to the second, the third to the third, the fourth to the fourth, then the fifth image file to the first script file. Or it could write the first 21000/4 image files to the first script file, then the second 21000/4 to the second, etc.

But that is a scripting task, irrelevant to ImageMagick, so I will leave that to you.
